File: Python/859. Buddy Strings.py
class Solution:
    def buddyStrings(self, s: str, goal: str) -> bool:
        if len(s) != len(goal): return False
        temp = []
        for i, j in enumerate(s):
            if goal[i] != j:
                temp.append(i)
        if len(temp) > 2: return False
        elif len(temp) == 2:
            goal = list(goal)
            goal[temp[0]], goal[temp[1]] = goal[temp[1]], goal[temp[0]]
            goal = "".join(goal)
            return s == goal
        else:
            if s == goal:
                return max(Counter(s).values()) >= 2
            else:
                return False

File: Rust/859. Buddy Strings.rs
use std::collections::HashMap;

impl Solution {
    pub fn buddy_strings(s: String, goal: String) -> bool {
        if s.len() != goal.len() { return false; }

        let mut temp: Vec<usize> = Vec::new();

        for (i, j) in s.chars().enumerate() {
            if let Some(t) = goal.chars().nth(i) {
                if j != t {
                    temp.push(i);
                }
            }
        }

        if temp.len() > 2 {
            return false;
        }
        else if temp.len() == 2 {
            return goal.chars().nth(temp[0]) == s.chars().nth(temp[1]) &&
                   goal.chars().nth(temp[1]) == s.chars().nth(temp[0]);
        }
        else {
            if s == goal {
                let mut temp: HashMap<char, i32> = HashMap::new();

                for i in s.chars() {
                    if temp.contains_key(&i) {
                        *temp.get_mut(&i).unwrap() += 1;
                    }
                    else {
                        temp.insert(i, 1);
                    }
                }

                for i in temp.values() {
                    if i >= &2 {
                        return true;
                    }
                }

                return false;
            }
            else {return false;}
        }
    }
}
